### PR TITLE
PRMT-3528 - Refactor defaultresult in EHR Out Service

### PR DESCRIPTION
**Pull request by joefong-nhs**
- Remove the usage of the mutable `defaultResult` object in `transferOutEhrCore`, replacing it with explicit logging and error handling
- Remove the `doTransfer` variable which served no real purpose and impeded code readability 
- Refactor `transferOutEhrCore`, group related logic into into helper functions for better readability

### CODE REVIEW

File: src/services/transfer/transfer-out-ehr-core.js [hunk 1]
```diff
@@ -18,65 +18,86 @@ export async function transferOutEhrCore({ conversationId, nhsNumber, messageId,
...
+    logInfo(`Replaced fragment id references in ehrCore`);
+  }
+  return { ehrCoreWithUpdatedMessageId, newMessageId };
+}
+
+const isEhrRequestDuplicate = async (conversationId) => {
+  const previousTransferOut  = await getRegistrationRequestStatusByConversationId(conversationId);
+  if (previousTransferOut !== null) {
+    logInfo('Duplicate transfer out request');
+    logInfo('EHR out transfer with this conversation ID is already in progress');
```
Comment (martin-nhs) [inline src/services/transfer/transfer-out-ehr-core.js:99]: I recon these logs both say the same thing, I would remove one of them and keep the other :)

```suggestion
    logInfo(`EHR out transfer with conversation ID ${conversationId} is already in progress`);
```

File: src/services/transfer/transfer-out-ehr-core.js [hunk 2]
```diff
@@ -18,65 +18,86 @@ export async function transferOutEhrCore({ conversationId, nhsNumber, messageId,
...
-      logInfo(`Successfully created new message id for all fragments`);
-      ehrCoreWithUpdatedMessageId = await updateReferencedFragmentIds(ehrCoreWithUpdatedMessageId);
-      logInfo(`Successfully replaced fragment id reference in ehrCore`);
-    }
+    const { ehrCoreWithUpdatedMessageId, newMessageId } = await getEhrCoreAndUpdateMessageIds(
+      nhsNumber,
+      conversationId
+    );
+    // TODO: Here we should update the message id of this Registration Request in database to newMessageId.
+    // To be addressed in another ticket
```
Comment (martin-nhs) [inline src/services/transfer/transfer-out-ehr-core.js:48]: Can we remove this TODO please, if no action has already been taken can we pop this in another ticker for the backlog? :D

File: src/services/transfer/__tests__/transfer-out-ehr-core.test.js [hunk 3]
```diff
@@ -52,60 +51,58 @@ describe('transferOutEhrCore', () => {
     payload: 'payload XML with updated referenced fragment message ids'
   };
 
-  patientAndPracticeOdsCodesMatch.mockResolvedValue(true);
 
-  updateMessageIdForEhrCore.mockImplementation(ehrCore => ehrCoreWithUpdatedMessageId);
+  afterEach(() => {
+    jest.resetAllMocks();
+  })
```
Comment (martin-nhs) [inline src/services/transfer/__tests__/transfer-out-ehr-core.test.js:57]: Super small optional one, can we slap a semicolon on the end of this please? :D

```suggestion
  afterEach(() => {
    jest.resetAllMocks();
  });
```

File: src/services/transfer/__tests__/transfer-out-ehr-core.test.js [hunk 4]
```diff
@@ -52,60 +51,58 @@ describe('transferOutEhrCore', () => {
...
       expect(getEhrCoreAndFragmentIdsFromRepo).toHaveBeenCalledWith(nhsNumber, conversationId);
-      expect(updateConversationStatus).toHaveBeenCalledWith(
-        conversationId,
-        Status.EHR_DOWNLOAD_FAILED
-      );
+      expect(updateConversationStatus).toHaveBeenCalledWith(conversationId, Status.EHR_DOWNLOAD_FAILED);
       expect(logInfo).toHaveBeenCalledWith(`Getting patient health record from EHR repo`);
+      expect(logError).toHaveBeenCalledWith('EHR transfer out request failed', new DownloadError());
       expect(sendCore).not.toHaveBeenCalled();
     });
```
Comment (martin-nhs) [inline src/services/transfer/__tests__/transfer-out-ehr-core.test.js:107]: Similar to above, can we split this test up with given, when and then comments please? :)

File: src/services/transfer/__tests__/transfer-out-ehr-core.test.js [hunk 5]
```diff
@@ -198,58 +178,54 @@ describe('transferOutEhrCore', () => {
...
+    const error = new Error('test error message');
     getRegistrationRequestStatusByConversationId.mockRejectedValueOnce(error);
 
-    const result = await transferOutEhrCore({ conversationId, nhsNumber, messageId, odsCode, ehrRequestId });
-    expect(result.hasFailed).toBe(true);
-    expect(result.error).toBe('test error message');
+    await transferOutEhrCore({ conversationId, nhsNumber, messageId, odsCode, ehrRequestId });
     expect(logError).toHaveBeenCalledWith('EHR transfer out request failed', error);
     expect(sendCore).not.toHaveBeenCalled();
   });
```
Comment (martin-nhs) [inline src/services/transfer/__tests__/transfer-out-ehr-core.test.js:193]: Similar to above can we split this test up with given, when and then comments please? :)

Comment (joefong-nhs) [reply src/services/transfer/__tests__/transfer-out-ehr-core.test.js]: Thank you so much!! I have address the PR comments :D

**Review by MohammadIqbalAD-NHS — APPROVED** [review]

**Pull request merged**